pyopencl/array.py

```diff
@@ -705,6 +736,63 @@ def __init__(
                          "than expected, potentially leading to crashes.",
                          InconsistentOpenCLQueueWarning, stacklevel=2)
 
+    # {{{ Pickling
+
+    def __getstate__(self):
+        try:
+            queue = _QUEUE_FOR_PICKLING_TLS.queue
+        except AttributeError:
+            queue = None
+
+        if queue is None:
+            raise RuntimeError("CL Array instances can only be pickled while "
+                               "queue_for_pickling is active.")
+
+        d = {}
+        d["shape"] = self.shape
+        d["dtype"] = self.dtype
+        d["strides"] = self.strides
+        d["allocator"] = self.allocator
+        d["nbytes"] = self.nbytes
+        d["strides"] = self.strides
+        d["offset"] = self.offset
+        d["data"] = self.get(queue=queue)
+        d["_flags"] = self._flags
+        d["size"] = self.size
+
+        return d
+
+    def __setstate__(self, state):
+        try:
+            queue = _QUEUE_FOR_PICKLING_TLS.queue
+        except AttributeError:
+            queue = None
+
+        if queue is None:
+            raise RuntimeError("CL Array instances can only be pickled while "
+                               "queue_for_pickling is active.")
+
+        self.queue = queue
+        self.shape = state["shape"]
+        self.dtype = state["dtype"]
+        self.strides = state["strides"]
+        self.allocator = state["allocator"]
+        self.offset = state["offset"]
+        self._flags = state["_flags"]
+        self.size = state["size"]
+        self.nbytes = state["nbytes"]
+        self.events = []
+
+        if self.allocator is None:
+            context = queue.context
+            self.base_data = cl.Buffer(context, cl.mem_flags.READ_WRITE, self.nbytes)
+        else:
+            self.base_data = self.allocator(self.nbytes)
```

Comment:
It looks like this would allocate redundantly if there were multiple arrays using the same buffer.

Reply:
Hmm, do you have a suggestion on how to handle this case?

Reply:
`b = a[:]` will make another array using the same buffer. (Any slice really.)

A way around this would be to make buffers picklable instead.

Reply:
Is [490bb94](https://github.com/inducer/pyopencl/pull/786/commits/490bb945a387e91ff3c59fdf012ee34aa48ee08f) going in the direction you had in mind? Should this be a separate PR?

Reply:
Yes, it does. I'm just not clear on why we need two separate `queue_for_pickling` decorators.

Reply:
> I'm just not clear on why we need two separate `queue_for_pickling` decorators.

60e71eeb935740205c4e4a2c1bb0687f4e71574b removes the duplicate.